File: backend/app/services/judge0_service.py
import time
import requests
import os
from app.config.settings import settings
from app.utils.exceptions import Judge0UnavailableException

class Judge0Service:
    def __init__(self, base_url: str):
        self.base_url = base_url.rstrip('/')

    def submit_code(self, source_code: str, language_id: int, stdin: str = "") -> str:
        url = f"{self.base_url}/submissions?base64_encoded=false"
        payload = {
            "source_code": source_code,
            "language_id": language_id,
            "stdin": stdin
        }
        try:
            response = requests.post(url, json=payload, timeout=10)
            response.raise_for_status()
            return response.json()["token"]
        except Exception as e:
            raise Judge0UnavailableException(f"Failed to submit code to Judge0: {str(e)}")

    def get_submission_result(self, token: str) -> dict:
        url = f"{self.base_url}/submissions/{token}?base64_encoded=false"
        try:
            response = requests.get(url, timeout=10)
            response.raise_for_status()
            data = response.json()
            return {
                "stdout": data.get("stdout") or "",
                "stderr": data.get("stderr") or "",
                "compile_output": data.get("compile_output") or "",
                "status": data.get("status") or {"id": 13, "description": "Internal Error"},
                "time": data.get("time") or "0.0",
                "memory": data.get("memory") or 0
            }
        except Exception as e:
            raise Judge0UnavailableException(f"Failed to get submission result from Judge0: {str(e)}")

    def execute_code(self, source_code: str, language_id: int, stdin: str = "") -> dict:
        token = self.submit_code(source_code, language_id, stdin)
        # Poll until finished
        max_attempts = 20
        for _ in range(max_attempts):
            result = self.get_submission_result(token)
            status_id = result["status"]["id"]
            if status_id not in [1, 2]: # 1: In Queue, 2: Processing
                return result
            time.sleep(0.5)
        
        return {
            "stdout": "",
            "stderr": "Execution timed out",
            "compile_output": "",
            "status": {"id": 5, "description": "Time Limit Exceeded"},
            "time": "0.0",
            "memory": 0
        }

class MockJudge0Service:
    def submit_code(self, source_code: str, language_id: int, stdin: str = "") -> str:
        return "mock-token-12345"

    def get_submission_result(self, token: str) -> dict:
        return {
            "stdout": "hello\n" if "hello" in token else "mock output\n",
            "stderr": "",
            "compile_output": "",
            "status": {"id": 3, "description": "Accepted"},
            "time": "0.01",
            "memory": 10240
        }

    def execute_code(self, source_code: str, language_id: int, stdin: str = "") -> dict:
        # Simple simulation based on source code content
        if "print('hello')" in source_code or 'print("hello")' in source_code:
            stdout = "hello\n"
        elif "print('Hello World')" in source_code or 'print("Hello World")' in source_code:
            stdout = "Hello World\n"
        else:
            stdout = "mock output\n"
            
        return {
            "stdout": stdout,
            "stderr": "",
            "compile_output": "",
            "status": {"id": 3, "description": "Accepted"},
            "time": "0.01",
            "memory": 10240
        }

def get_judge0_service():
    # If JUDGE0_URL environment variable is missing or empty, fallback to Mock
    env_url = os.getenv("JUDGE0_URL")
    if not env_url:
        return MockJudge0Service()
    return Judge0Service(settings.resolved_judge0_url)

File: backend/app/services/judge0_service.js
const axios = require('axios');
const settings = require('../config/settings');
const { Judge0UnavailableException } = require('../utils/exceptions');

class Judge0Service {
  constructor(baseUrl) {
    this.baseUrl = baseUrl.replace(/\/+$/, '');
  }

  async submitCode(sourceCode, languageId, stdin = "") {
    const url = `${this.baseUrl}/submissions?base64_encoded=false`;
    const payload = {
      source_code: sourceCode,
      language_id: languageId,
      stdin: stdin
    };
    try {
      const response = await axios.post(url, payload, { timeout: 10000 });
      return response.data.token;
    } catch (e) {
      throw new Judge0UnavailableException(`Failed to submit code to Judge0: ${e.message}`);
    }
  }

  async getSubmissionResult(token) {
    const url = `${this.baseUrl}/submissions/${token}?base64_encoded=false`;
    try {
      const response = await axios.get(url, { timeout: 10000 });
      const data = response.data;
      return {
        stdout: data.stdout || "",
        stderr: data.stderr || "",
        compile_output: data.compile_output || "",
        status: data.status || { id: 13, description: "Internal Error" },
        time: data.time || "0.0",
        memory: data.memory || 0
      };
    } catch (e) {
      throw new Judge0UnavailableException(`Failed to get submission result from Judge0: ${e.message}`);
    }
  }

  async executeCode(sourceCode, languageId, stdin = "") {
    const token = await this.submitCode(sourceCode, languageId, stdin);
    const maxAttempts = 20;
    for (let i = 0; i < maxAttempts; i++) {
      const result = await this.getSubmissionResult(token);
      const statusId = result.status.id;
      if (statusId !== 1 && statusId !== 2) {
        return result;
      }
      await new Promise(resolve => setTimeout(resolve, 500));
    }

    return {
      stdout: "",
      stderr: "Execution timed out",
      compile_output: "",
      status: { id: 5, description: "Time Limit Exceeded" },
      time: "0.0",
      memory: 0
    };
  }
}

class MockJudge0Service {
  async submitCode(sourceCode, languageId, stdin = "") {
    return "mock-token-12345";
  }

  async getSubmissionResult(token) {
    return {
      stdout: token.includes("hello") ? "hello\n" : "mock output\n",
      stderr: "",
      compile_output: "",
      status: { id: 3, description: "Accepted" },
      time: "0.01",
      memory: 10240
    };
  }

  async executeCode(sourceCode, languageId, stdin = "") {
    let stdout = "mock output\n";
    if (sourceCode.includes("print('hello')") || sourceCode.includes('print("hello")') || sourceCode.includes('console.log("hello")') || sourceCode.includes("console.log('hello')")) {
      stdout = "hello\n";
    } else if (sourceCode.includes("print('Hello World')") || sourceCode.includes('print("Hello World")')) {
      stdout = "Hello World\n";
    }
    return {
      stdout: stdout,
      stderr: "",
      compile_output: "",
      status: { id: 3, description: "Accepted" },
      time: "0.01",
      memory: 10240
    };
  }
}

function getJudge0Service() {
  const envUrl = process.env.JUDGE0_URL;
  if (envUrl === undefined || envUrl === "") {
    return new MockJudge0Service();
  }
  return new Judge0Service(settings.resolved_judge0_url);
}

module.exports = {
  getJudge0Service,
  Judge0Service,
  MockJudge0Service
};
